### PR TITLE
add dgram unix socket channel

### PR DESCRIPTION
**Pull request by yuanwang-wf**
Issue https://github.com/jnr/jnr-unixsocket/issues/22

### CODE REVIEW

File: src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+  /**
+   * invoke native socketpair.
+   *
+   * @return an array of two DgramUnixSocketChannels
+   * @throws IOException if socket fails to connect
+   */
+  public static final DgramUnixSocketChannel[] pair() throws IOException {
+    int[] sockets = {-1, -1};
+    Native.socketpair(ProtocolFamily.PF_UNIX, Sock.SOCK_DGRAM, 0, sockets);
```
Comment (yuanwang-wf) [inline src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java]: this is basically the only difference between this and ```UnixSocketChannel```, I can add an abstract SocketChannel to capture the shared logic.

File: src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+      return true;
+    }
+  }
+
+  /**
+   * check whether current socket is connected.
+   *
+   * @return whether current connection is connected
+   */
+  public boolean isConnected() {
```
Comment (yuanwang-wf) [inline src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java]: i am not sure whether the following methods apply to DGRAM socket or not

File: src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java (hunk 1 above)
Comment (headius) [inline src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java]: Please do add an abstract superclass they can both share, please.

File: src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java (hunk 2 above)
Comment (headius) [inline src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java]: They should probably just return sensible defaults for a dgram (presumably connectionless) socket.

File: src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java (hunk 1 above)
Comment (yuanwang-wf) [inline src/main/java/jnr/unixsocket/DgramUnixSocketChannel.java]: will do

**Comment by felfert** [comment]
I'm afraid it is not **that** simple. A proper implementation would do the following:

1. start in jnr/jnr-enxio with some refactoring:
  NativeSocketChannel is quite SOCK_STREAMish, so I would add a new class NativeDatagramChannel with the following changes:
 - No shutdownInput() and shutdownOutput(), since those are SOCK_STREAM only.
 - Implement receive() like in [java.nio.channels.DatagramChannel](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/DatagramChannel.html#receive(java.nio.ByteBuffer))
 - Implement send() like in [java.nio.channels.DatagramChannel](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/DatagramChannel.html#send(java.nio.ByteBuffer,%20java.net.SocketAddress))


2. Then, in order to reflect the connectionless semantic, the new class should be modeled after java.nio.channels.DatagramChannel (just like UnixSocketChannel is modeled after  java.nio.channels.SocketChannel),  and extend the new NativeDatagramChannel from jnr-enxio. Particular changes:
 - open() should **not** call connect()
 - remove doConnect(), finishConnect(), connect() sets state only.
 - add disconnect() like connect() basically a NOP except for setting state of course.
 - add setOption()
 - add wrappers for read() and write() which throw an exception, if state is not connected.

It just so happens, that I just now need a SOCK_DGRAM implementation as well :-) (for writing to /dev/log), so I guess I'll start with 1.) ...
@yuanwang-wf Perhaps we can collaborate. If you like, just contact me via mail.

**Comment by felfert** [comment]
@headius Your travis setup seems to be broken in that it's maven build still attempts to use the obsolete codehaus repo. Can you fix that please?

**Comment by felfert** [comment]
Now collaborating with marcuslinke in his PR #35 (WIP, but already functional)

**Pull request closed without merging**